### ISSUE TITLE
Loop Strength Reduction

### ISSUE THREAD

**Issue by dubiousconst282** [dubiousconst282/DistIL#12]
In loops over sequential memory locations (arrays, spans, lists), it may be more efficient to use the base pointer as the induction variable, rather than indexing on every iteration.

```cs
//Original
int[] array = ...
for (int i = 0; i < array.Length; i++) {
    Use(array[i]); //JIT currently emits something like `mov temp, dword[array + i*4 + 0x10]`, plus maybe a sign extension for i
}
//Possible transform
int& ptr = MemoryMarshal.GetArrayDataReference(array);
int& endPtr = ptr + array.Length;
for (; ptr < endPtr; ptr++) {
    Use(*ptr); //should result into something like `mov temp, dword[ptr]`
}
```

The linq expansion pass already does this for arrays and lists.

Note that this can only be safely done if the loop bound is known to be less than the array length (due to bound checks), which may only be unblocked with something like value range analysis.

It is a relatively dangerous transform - it could inadvertently take away opportunities from JIT's own optimizations (which?), increase register pressure (if not all IVs and the loop condition are replaced), and add loop-carried dependencies ([ref](https://stackoverflow.com/questions/72306573/why-does-this-code-execute-more-slowly-after-strength-reducing-multiplications-t)).

The main benefit may actually be simplification - removing array bound checks and canonicalizing accesses to LoadPtrInst may make it easier for the future vectorization pass.

Since the impact is otherwise relatively small and there are plans for it to be implemented in the JIT (ref: https://github.com/dotnet/runtime/issues/65342), it may not be worth spending too much time on it.

TODOs:
- [x] Recognize induction variables (basic and derived)
   - _bin ops only supports constants (e.g. `derivIV = mul baseIV, 4`); support for loop invariants would require a more complicated IV representation, for probably little benefit_
- Rewriting:
  - [x] Support for arrays
  - ~Support bin ops (multiply)~
     - Not always beneficial
  - [x] Support for pointers (recognize derived IVs based on conv instruction)
- [ ] Support for spans (recognize and rewrite)
- [ ] More testing

**Comment by dubiousconst282** [comment]
Pointer index reduction is now supported for simple cases with primitive types (having known sizes), and it is limited for cases where the base address has a single use inside the loop (the candidate IV), example:

```cs
public static float SumAbsDiffsPtr(float* a, float* b, int count) {
    float sum = 0;
    for (int i = 0; i < count; i++) {
        sum += Math.Abs(a[i] - b[i]);
    }
    return sum;
}

//Optimized IR (forest dump)
BB_01:
  r4 = add #a, (mul (conv #count), 4) -> float*
  goto BB_16
BB_06:
  r11 = fadd r17, (call Math::Abs(float: (fsub (ldptr lsr_ptr), (ldptr lsr_ptr2)))) -> float
  r12 = add r18, 1 -> int
  r13 = add lsr_ptr, 4 -> float*
  r14 = add lsr_ptr2, 4 -> float*
  goto BB_16
BB_16:
  r17 = phi [BB_01: 0.0f], [BB_06: r11] -> float
  r18 = phi [BB_01: 0], [BB_06: r12] -> int
  lsr_ptr = phi [BB_01: #a], [BB_06: r13] -> float*
  lsr_ptr2 = phi [BB_01: #b], [BB_06: r14] -> float*
  goto (icmp.ult lsr_ptr, r4) ? BB_06 : BB_23
BB_23:
  ret r17
```

Dead phis should ideally be eliminated by DCE in the future.

The limitation around pointers with single-uses mainly comes from the lack of support for non-constant offsets in the derived IV representation, though we'd also need to find the "best IV" to avoid reducing too many variables and increasing register pressure. A sample where this issue would happen:

```cs
public static void ConvertRgbaToYuva(byte* src, byte* dest, int count) {
    for (int i = 0; i < count; i++) {
        byte r = src[i * 4 + 0]; //only this access would be reduced
        byte g = src[i * 4 + 1];
        byte b = src[i * 4 + 2];
        byte a = src[i * 4 + 3];
        dest[i * 4 + 0] = (byte)(  0 + ((r *  77 + g *  150 + b *  29) >> 8));
        dest[i * 4 + 1] = (byte)(128 + ((r * -43 + g *  -85 + b * 128) >> 8));
        dest[i * 4 + 2] = (byte)(128 + ((r * 128 + g * -107 + b * -21) >> 8));
        dest[i * 4 + 3] = a;
    }
}

//Forest dump
BB_01:
  goto BB_49
BB_03:
  r4 = mul r50, 4 -> int
  r6 = ldptr (add #src, r4) -> byte     //DIV recognized as:  r50 * 4 + #src
  r7 = add r4, 1 -> int                 //Ignored (not pointer-based)
  r9 = ldptr (add #src, r7) -> byte     //DIV not recognized, conflicting non-constant offsets (1, #src)
  r10 = add r4, 2 -> int
  r12 = ldptr (add #src, r10) -> byte
  r13 = add r4, 3 -> int
  r15 = ldptr (add #src, r13) -> byte
  stptr (add #dest, r4), (conv (shra (add (add (mul r6, 77), (mul r9, 150)), (mul r12, 29)), 8)) as sbyte
  stptr (add #dest, r7), (conv (add (shra (add (add (mul r6, -43), (mul r9, -85)), (mul r12, 128)), 8), 128)) as sbyte
  stptr (add #dest, r10), (conv (add (shra (add (add (mul r6, 128), (mul r9, -107)), (mul r12, -21)), 8), 128)) as sbyte
  stptr (add #dest, r13), r15 as sbyte
  r47 = add r50, 1 -> int
  goto BB_49
BB_49:
  r50 = phi [BB_01: 0], [BB_03: r47] -> int
  goto (icmp.slt r50, #count) ? BB_03 : BB_53
BB_53:
  ret

//Reduced (no single-use heuristic)
BB_01:
  r3 = add #src, (conv #count) -> byte*
  goto BB_51
BB_05:
  r6 = mul r52, 4 -> int
  r7 = ldptr lsr_ptr -> byte
  r8 = add r6, 1 -> int
  r10 = ldptr (add #src, r8) -> byte
  r11 = add r6, 2 -> int
  r13 = ldptr (add #src, r11) -> byte
  r14 = add r6, 3 -> int
  r16 = ldptr (add #src, r14) -> byte
  stptr lsr_ptr2, (conv (shra (add (add (mul r7, 77), (mul r10, 150)), (mul r13, 29)), 8)) as sbyte
  stptr (add #dest, r8), (conv (add (shra (add (add (mul r7, -43), (mul r10, -85)), (mul r13, 128)), 8), 128)) as sbyte
  stptr (add #dest, r11), (conv (add (shra (add (add (mul r7, 128), (mul r10, -107)), (mul r13, -21)), 8), 128)) as sbyte
  stptr (add #dest, r14), r16 as sbyte
  r47 = add r52, 1 -> int
  r48 = add lsr_ptr, 4 -> byte*
  r49 = add lsr_ptr2, 4 -> byte*
  goto BB_51
BB_51:
  r52 = phi [BB_01: 0], [BB_05: r47] -> int
  lsr_ptr = phi [BB_01: #src], [BB_05: r48] -> byte*
  lsr_ptr2 = phi [BB_01: #dest], [BB_05: r49] -> byte*
  goto (icmp.ult lsr_ptr, r3) ? BB_05 : BB_57
BB_57:
  ret
```

**Comment by dubiousconst282** [comment]
Closing in favor of a simpler approach which only works on for-i loops with arrays (3d6d63f62).
The existing code was overly complicated because it tried to decompose accesses based on basic and derived IVs (somewhat like SCEVs but worse), which also made it harder to tune.

There wasn't really a reason for it to be that complex since indexing is quite cheap anyway. Also, out of bounds GC pointers are quite problematic so it wouldn't be possible to reduce something like (i * 2 - 5) unless we knew for sure it was in bounds.
